### PR TITLE
Fix the bugs in Tx & PI circuits reported by Zellic & KALOS auditors

### PR DESCRIPTION
**Pull request by kunxian-xia**
### Description

This PR aims to fix all the bugs of tx circuit reported in the [Zellic/Kalos Audit Report](https://hackmd.io/4XuxkwHERa6AIPQ6tr2QAg).

### Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

### Contents

- [x] Finding 19
- [x] https://github.com/scroll-tech/zkevm-circuits/pull/613
- [x] https://github.com/scroll-tech/zkevm-circuits/pull/614
- [x] https://github.com/scroll-tech/zkevm-circuits/pull/618
- [x] https://github.com/scroll-tech/zkevm-circuits/pull/656
- [x] https://github.com/scroll-tech/zkevm-circuits/pull/619
- [x] https://github.com/scroll-tech/zkevm-circuits/pull/622
- [x] https://github.com/scroll-tech/zkevm-circuits/pull/735
- [x] https://github.com/scroll-tech/zkevm-circuits/pull/624
- [x] https://github.com/scroll-tech/zkevm-circuits/pull/623
- [x] https://github.com/scroll-tech/zkevm-circuits/pull/733

### CODE REVIEW

File: zkevm-circuits/src/tx_circuit.rs [hunk 1]
```diff
@@ -772,16 +820,49 @@ impl<F: Field> SubCircuitConfig<F> for TxCircuitConfig<F> {
                     meta.query_advice(calldata_gas_cost_acc, Rotation::next()),
                     meta.query_advice(calldata_gas_cost_acc, Rotation::cur()) + gas_cost_next,
                 );
+                cb.require_equal(
+                    "calldata_rlc' = calldata_rlc * r + byte'",
+                    meta.query_advice(calldata_rlc, Rotation::next()),
+                    meta.query_advice(calldata_rlc, Rotation::cur()) * challenges.keccak_input()
+                        + meta.query_advice(tx_table.value, Rotation::next()),
```
Comment (sampritipanda) [inline zkevm-circuits/src/tx_circuit.rs:962]: This calculation is insecure because `tx_table.value` is in SecondPhase.

Comment (kunxian-xia) [reply zkevm-circuits/src/tx_circuit.rs:962]: fixed in the https://github.com/scroll-tech/zkevm-circuits/pull/735. I copied the tx_table.value of calldata part to a newly allocated 1st phase column.

File: zkevm-circuits/src/tx_circuit.rs [hunk 2]
```diff
@@ -695,6 +754,44 @@ impl<F: Field> SubCircuitConfig<F> for TxCircuitConfig<F> {
...
+                let cum_num_txs = meta.query_advice(cum_num_txs, Rotation::cur());
+
+                cum_num_txs - num_txs
+            },
+            |meta| meta.query_advice(tx_table.tx_id, Rotation::cur()),
+            u8_table.into(),
+        );
+
+        // last non-padding tx must have tx_id == cum_num_txs
+        meta.create_gate(
```
Comment (lispc) [inline zkevm-circuits/src/tx_circuit.rs:780]: will these 2 gates still ok since we need to handle the l1 tx queue index design now?

**Comment by lispc** [comment]
2 changes needs to be considered carefully

(1) NumTx: the complex l1 tx num mechanism
(2) TxFieldTag::TxType: recently to handle l1 tx, we added this to tx table. This may affect some hardcode offset (if any)

File: zkevm-circuits/src/tx_circuit.rs (hunk 2 above)
Comment (kunxian-xia) [reply zkevm-circuits/src/tx_circuit.rs:780]: yes, the l1 queue index is only relevant to `num_all_txs`.

**Pull request merged**